src/python/pants/backend/python/lint/black/rules.py

```diff
@@ -49,20 +49,18 @@ async def black_fmt(request: BlackRequest, black: Black, python_setup: PythonSet
     # when relevant. We only do this if if <3.8 can't be used, as we don't want a loose requirement
     # like `>=3.6` to result in requiring Python 3.8, which would error if 3.8 is not installed on
     # the machine.
-    all_interpreter_constraints = InterpreterConstraints.create_from_compatibility_fields(
-        (field_set.interpreter_constraints for field_set in request.field_sets),
-        python_setup,
-    )
-    tool_interpreter_constraints = (
-        all_interpreter_constraints
-        if (
-            black.options.is_default("interpreter_constraints")
-            and all_interpreter_constraints.requires_python38_or_newer(
-                python_setup.interpreter_universe
-            )
-        )
-        else black.interpreter_constraints
-    )
+    tool_interpreter_constraints = black.interpreter_constraints
+    if black.options.is_default("interpreter_constraints"):
+        distinct_constraints = InterpreterConstraints.group_field_sets_by_constraints(
+            request.field_sets, python_setup
+        ).keys()
+        for constraints in distinct_constraints:
+            if constraints.requires_python38_or_newer(python_setup.interpreter_universe):
+                # We've found at least one constraint that requires 3.8, so we know we must run
+                # black on 3.8+, and this constraint is as good as any (in the sense that we
+                # expect a compatible interpreter to exist).
+                tool_interpreter_constraints = constraints
+                break
```

Comment:
This is not sufficient. We have to loop all the way through in case there is 3.9+ vs 3.8+. We must use the highest bound

Reply:
I've reverted this whole thing in favor of the Alexandrian stroke of a better error message...

Reply:
Alexandrian stroke?

Reply:
Refers to the [Gordian knot](https://en.wikipedia.org/wiki/Gordian_Knot)